woocommerce/google-listings-and-ads
# Get suggested shipping rates for provinces/states and postal codes from WC zones

### Changes proposed in this Pull Request:



Closes #1230

This PR introduces new services to parse the WooCommerce shipping zones and return the shipping rates for each country and its subdivisions (state, provides, postcodes, etc.).

The `ShippingZone` service functionality is divided into three other classes that handle parsing the locations, methods, and grouping them together.

Instead of mapping the rates as multidimensional arrays, several value classes are introduced to hold `Location`, different types of `ShippingRate`, and a combination of the two: `LocationRate`. This allows us to have a consistent verifiable format when dealing with the suggested shipping rates throughout the codebase.

The `ShippingZone` service has now a new function that returns the rates grouped by countries. This is used in the suggestions API since we currently only suggest rates on a country level.

The previous method `get_shipping_rates_for_country` is now refactored to return the shipping rates for all subdivisions of a country as well.

All shipping rate suggestions are returned as a `LocationRate` object.

I've also removed the `currency` property from suggested shipping rates since they all always have the same currency as WooCommerce. The `currency` for the rates is now directly fetched from WooCommerce in suggestions API and Googe settings sync service.


### Detailed test instructions:


The new functionality to return the shipping rates for subdivisions isn't being used anywhere in the code so it's not possible to test it without manually calling the `ShippingZone::get_shipping_rates_for_country` method somewhere and inspecting its output.

1. Define multiple shipping zones each containing various regions, postcodes, shipping methods, and rates for shipping classes.
2. Make a GET request to the `mc/shipping/rates/suggestions` API with the country_codes parameter set to an array of country codes from your defined shipping zones and confirm the rates are returned the same as you defined in your shipping zone settings. Note that the rates will be grouped at a country level.
3. Send a `POST` request to `wc/gla/mc/settings` and include the following data as JSON body: `{ "shipping_rate":"automatic" }`
4. Send a `POST` request to `wc/gla/mc/settings/sync`
5. Confirm that it's successful.
6. Check the shipping rates in your MC account and confirm that they match what you defined in Woo shipping zones. Again, the syncing functionality is still working on a country level so you won't see any state-specific rates synced.

To test that the code also returns rates on a subdivision level, you can call the `ShippingZone::get_shipping_rates_for_country` method anywhere in the code (for example, in the shipping rates suggestions controller) and inspect its output via xDebug or simply `var_dump`. It will return the rates as a flat array but each will have its `location` property set correctly.


### Additional details:

A side effect of these changes is that the `GET mc/shipping/rates/suggestions` API now returns the suggestions differently. Previously, if a `free_shipping` rate with a `free_shipping_threshold` was among the suggestions, this API would return it as a `flat_rate` with a `rate` of `0`. With this PR, the `free_shipping` suggestion will be returned as-is.

The other important change is that the `free_shipping_threshold` is now only expected to be set as an `option` for the `free_shipping` method. A `flat_rate` method no longer should have a `free_shipping_threshold` option.

For example, if the following shipping zones were created in WC settings:
![image](https://user-images.githubusercontent.com/73110514/160553823-0742ef36-0bc3-47ce-9325-2e594021fef2.png)

This would be the response of `GET mc/shipping/rates/suggestions`:
```JSON
[
    {
        "country": "TR",
        "method": "free_shipping",
        "currency": "USD",
        "rate": 0,
        "options": {}
    },
    {
        "country": "US",
        "method": "flat_rate",
        "currency": "USD",
        "rate": 20,
        "options": {}
    },
    {
        "country": "US",
        "method": "free_shipping",
        "currency": "USD",
        "rate": 0,
        "options": {
            "free_shipping_threshold": 4.99
        }
    },
    {
        "country": "ES",
        "method": "free_shipping",
        "currency": "USD",
        "rate": 0,
        "options": {
            "free_shipping_threshold": 100
        }
    }
]
```

Note the suggested shipping rates for US. Previously the "free_shipping_threshold" would be merged with the "flat_rate" method but now it's on a separate "free_shipping" rate.

The `POST mc/shipping/rates` API is also affected in the same way. This means that the `free_shipping_threshold` option is always expected to be set on a `free_shipping` rate and Free Shipping should be defined as a rate with a type equal to `free_shipping`. A flat_rate with a cost of `0` would still work but it cannot have the `free_shipping_threshold` option set on it.

For instance, to set a flat rate of 20 with conditional free shipping of orders above 4.99 for the US, the following rates should be sent to the `POST mc/shipping/rates/batch` API:

```JSON
{
  "rates": [
    {
        "country": "US",
        "method": "flat_rate",
        "currency": "USD",
        "rate": 20,
        "options": {}
    },
    {
        "country": "US",
        "method": "free_shipping",
        "currency": "USD",
        "rate": 0,
        "options": {
            "free_shipping_threshold": 4.99
        }
    }
  ]
}
```


### Changelog entry

> Add - Get shipping rates suggestions for provinces/states and postal codes
